### ISSUE TITLE
ICE: Encountered error `Unimplemented` when trying to select an implementation for constant trait item reference. <anon>:4     const ID: i32;

### ISSUE THREAD

**Issue by Mr-Byte** [rust-lang/rust#25606]
I have two similar pieces of code that I would expect to behave in a similar fashion.  The first piece of code compiles and functions properly:

```rust
trait HasId {
    fn id() -> i32;
}

trait Foo {}

impl HasId for Foo {
    fn id() -> i32 { 1 }
}

trait Bar {}

impl HasId for Bar {
    fn id() -> i32 { 2 }
}

fn print_id<T: HasId + ?Sized>() {
    println!("{}", <T as HasId>::id());
}

fn main() {
    print_id::<Foo>();
    print_id::<Bar>();
}
```

The second piece of code replaces the static methods with an associated constant:

```rust
#![feature(associated_consts)]

trait HasId {
    const ID: i32;
}

trait Foo {}

impl HasId for Foo {
    const ID: i32 = 1;
}

trait Bar {}

impl HasId for Bar {
    const ID: i32 = 2;
}

fn print_id<T: HasId + ?Sized>() {
    println!("{}", <T as HasId>::ID);
}

fn main() {
    print_id::<Foo>();
    print_id::<Bar>();
}
```

This code fails to compile with the following ICE:

```
<anon>:4:5: 4:19 error: internal compiler error: Encountered error `Unimplemented` when trying to select an implementation for constant trait item reference.
<anon>:4     const ID: i32;
             ^~~~~~~~~~~~~~
note: the compiler unexpectedly panicked. this is a bug.
note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports
note: run with `RUST_BACKTRACE=1` for a backtrace
thread 'rustc' panicked at 'Box<Any>', /home/rustbuild/src/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/libsyntax/diagnostic.rs:170
```

What's interesting is that the following code does compile:

```rust
#![feature(associated_consts)]

trait HasId {
    const ID: i32;
}

trait Foo {}

impl HasId for Foo {
    const ID: i32 = 1;
}

trait Bar {}

impl HasId for Bar {
    const ID: i32 = 2;
}

fn main() {
    println!("{}", <Foo as HasId>::ID);
    println!("{}", <Bar as HasId>::ID);
}
```

So it would appear as if this behavior is related to the use of a generic in the UFCS resolution for the associated constant.

**Comment by arielb1** [comment]
cc @quantheory

**Comment by arielb1** [comment]
I think you wrote your static-method example twice (I couldn't find the ICE-ing example).

**Comment by arielb1** [comment]
For the record, this ICEs:

```Rust
#![feature(associated_consts)]

trait HasId {
    const ID: i32;
}

fn get_id<T: HasId + ?Sized>() -> i32 {
    <T as HasId>::ID
}

fn main() {
}
```

**Comment by Mr-Byte** [comment]
Sorry about that, grabbed the same code twice.  I've updated my example that ICEs.  You were able to further reduce it down though, thanks!

**Comment by quantheory** [comment]
This is a duplicate of #25046 (see my comment there, and will be fixed by #25091. I would very much like to have this code actually work in the future, but for now the correct behavior is to reject such uses of associated constants.

**Comment by Aatch** [comment]
Duplicate of #25046
